prime-router/src/main/kotlin/fhirengine/engine/FHIRRouter.kt

```diff
@@ -183,21 +198,139 @@ class FHIRRouter(
+     * @return list of receivers that should receive this bundle
+     */
+    internal fun applyFilters(bundle: Bundle): List<Receiver> {
+        val listOfReceivers = mutableListOf<Receiver>()
+        // find all receivers that have the full ELR topic and determine which applies
+        val fullElrReceivers = settings.receivers.filter {
+            it.customerStatus != CustomerStatus.INACTIVE &&
+                it.topic == Topic.FULL_ELR.json_val
+        }
+
+        // get the quality filter default result for the bundle, but only if it is needed
+        val qualFilterDefaultResult: Boolean by lazy {
+            evaluateFilterCondition(
+                qualityFilterDefault, bundle,
+                false
+            )
+        }
+        // get the processing mode (processing id) default result for the bundle, but only if it is needed
+        val processingModeDefaultResult: Boolean by lazy {
+            evaluateFilterCondition(
+                processingModeFilterDefault, bundle,
+                false
+            )
+        }
+
+        fullElrReceivers.forEach { receiver ->
+            // get the receiver's organization, since we need to be able to find/combine the correct filters
+            val orgFilters = settings.findOrganization(receiver.organizationName)!!.filters
+
+            // Get the applicable filters, either receiver or organization level if there are no receiver filters
+
+            // NOTE: these could all be combined into a single `if` statement, but that would reduce readability and
+            // ability to debug this section so is being left split out like this
+
+            // JURIS FILTER
+            //  default: allowNone
+            var passes = evaluateFilterCondition(getJurisFilters(receiver, orgFilters), bundle, false)
+            // QUALITY FILTER
+            //  default: must have message id, patient last name, patient first name, dob, specimen type
+            //           must have at least one of patient street, zip code, phone number, email
+            //           must have at least one of order test date, specimen collection date/time, test result date
+            passes = passes &&
+                evaluateFilterCondition(getQualityFilters(receiver, orgFilters), bundle, qualFilterDefaultResult)
+            // ROUTING FILTER
+            //  default: allowAll
+            passes = passes &&
+                evaluateFilterCondition(getRoutingFilter(receiver, orgFilters), bundle, true)
+            // PROCESSING MODE FILTER
+            //  default: allowAll
+            passes = passes &&
+                evaluateFilterCondition(
+                    getProcessingModeFilter(receiver, orgFilters), bundle,
+                    processingModeDefaultResult
+                )
+
+            // if all filters pass, add this receiver to the list of valid receivers
+            if (passes) {
+                listOfReceivers.add(receiver)
+            }
```

Comment:
(Optional) Since we're breaking this down already, both readability and debug-ability could maybe be further enhanced by having separate variables for `jurisFilterPasses`, `qualityFilterPasses`, `routingFilterPasses`, and `modeFilterPasses`, and either `and`ing them at the end to evaluate `passes`, or making the "if all filters pass" condition be `jurisFilterPasses && qualityFilterPasses && routingFilterPasses && modeFilterPasses` and not using `passes` at all. Not familiar with logging norms, but that would also allow us to easily log which filter conditions passed or failed. Just an idea, feel free to keep it as it is.

Reply:
Originally I had it as you suggest, but in this case it is fall-through logic - this filters are in a particular order and if any filter higher up fails the ones below will not be run since it would short circuit past them. We don't want to run expensive complicated filters on a Bundle if it doesn't pass, for example, jurisdictional filter. In this code we are running ALL potential receivers against a bundle coming through so want to do the least amount of work necessary.

Reply:
Ah nice, I overlooked that short-circuiting, I like it.